Cabal/Distribution/Simple/PackageIndex.hs

```diff
@@ -138,7 +143,7 @@ data PackageIndex a = PackageIndex {
   --
   -- FIXME: Clarify what "preference order" means. Check that this invariant is
   -- preserved. See #1463 for discussion.
-  packageIdIndex :: !(Map PackageName (Map Version [a]))
+  packageIdIndex :: !(Map (PackageName, Maybe UnqualComponentName) (Map Version [a]))
```

Comment:
What about `Map PackageName (Map (Maybe UnqualComponentName) (Map Version [a]))`? This matches what you ended up on for `Backpack.ConfiguredComponent` (and what I did in my branch :D).